src/power_geopm/nrmpower_geopm.c

```diff
@@ -0,0 +1,377 @@
+/*******************************************************************************
+ * Copyright 2021 UChicago Argonne, LLC.
+ * (c.f. AUTHORS, LICENSE)
+ *
+ * This file is part of the nrm-extra project.
+ * For more info, see https://github.com/anlsys/nrm-extra
+ *
+ * SPDX-License-Identifier: BSD-3-Clause
+ *******************************************************************************/
+
+/* Filename: nrmpower_papi.c
+ *
+ * Description: Implements middleware between powercap, measured via PAPI,
+ *               and the NRM downstream interface. Resources detected via hwloc.
+ */
+
+#include <limits.h>
+#include <stddef.h>
+#define _GNU_SOURCE
+#include <assert.h>
+#include <errno.h>
+#include <geopm_pio.h>
+#include <geopm_topo.h>
+#include <getopt.h>
+#include <hwloc.h>
+#include <limits.h>
+#include <math.h>
+#include <sched.h>
+#include <signal.h>
+#include <stdarg.h>
+#include <stdbool.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+#include <sys/wait.h>
+#include <time.h>
+#include <unistd.h>
+
+#include <nrm.h>
+
+#include "extra.h"
+
+static int log_level = NRM_LOG_DEBUG;
+volatile sig_atomic_t stop;
+
+static nrm_client_t *client;
+static nrm_scope_t *scope;
+static nrm_sensor_t *sensor;
+
+static char *upstream_uri = "tcp://127.0.0.1";
+static int pub_port = 2345;
+static int rpc_port = 3456;
+
+#define assert__(x) for (; !(x); assert(x))
+
+char *usage =
+        "usage: nrm-power [options] \n"
+        "     options:\n"
+        "            -s	 --signals           Single GEOPM Signal name. Default: both CPU_POWER and DRAM_POWER\n"
+        "                                       Use `sudo geopmread` to determine valid signal names         \n"
+        "            -v, --verbose           Produce verbose output. Log messages will be displayed to stderr\n"
+        "            -h, --help              Displays this help message\n";
+
+// handler for interrupt?
+void interrupt(int signum)
+{
+	nrm_log_debug("Interrupt caught. Exiting loop.\n");
+	stop = 1;
+}
+
+double get_watts(double event_value, int64_t elapsed_time)
+{
+	return ((double)event_value / 1.0e6) / (elapsed_time / 1.0e9);
+}
+
+int get_cpu_idx(hwloc_topology_t topology, int cpu)
+{
+	hwloc_obj_t pu;
+	pu = hwloc_get_pu_obj_by_os_index(topology, cpu);
+	return pu->logical_index;
+}
+
+char *get_domain_label(char *string)
+{
+	char *label;
+	do {
+		label = strtok(string, "_");
+	} while (label != NULL);
+	return label; // should return last token after the last underscore
+}
+
+struct signal_info_s {
+	char *signal_name;
+	char *signal_token;
+	int domain_type;
+};
+
+typedef struct signal_info_s signal_info_t;
+
+int main(int argc, char **argv)
+{
+	int j, char_opt, err;
+	double freq = 1;
+	size_t i, MAX_SIGNAL_NAME_LENGTH = 55;
+	nrm_vector_t *signals = NULL;
+
+	// a vector of GEOPM signal names; will be pushed into by e.g.: -s
+	// DRAM_POWER -s CPU_POWER
+	assert(nrm_vector_create(&signals, MAX_SIGNAL_NAME_LENGTH) ==
+	       NRM_SUCCESS);
```

Comment:
you're pushing strings in this vector, the second argument should be sizeof(char *)